common/src/main/java/com/datastax/oss/sink/RecordProcessor.java

```diff
@@ -0,0 +1,268 @@
+/*
+ * Copyright DataStax, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.datastax.oss.sink;
+
+import com.datastax.oss.driver.api.core.cql.AsyncResultSet;
+import com.datastax.oss.driver.api.core.cql.BoundStatement;
+import com.datastax.oss.driver.shaded.guava.common.annotations.VisibleForTesting;
+import com.datastax.oss.driver.shaded.guava.common.util.concurrent.ThreadFactoryBuilder;
+import com.datastax.oss.sink.config.CassandraSinkConfig;
+import com.datastax.oss.sink.config.TableConfig;
+import com.datastax.oss.sink.config.TopicConfig;
+import com.datastax.oss.sink.metadata.InnerDataAndMetadata;
+import com.datastax.oss.sink.metadata.MetadataCreator;
+import com.datastax.oss.sink.record.HeadersDataMetadata;
+import com.datastax.oss.sink.record.KeyValueRecord;
+import com.datastax.oss.sink.record.KeyValueRecordMetadata;
+import com.datastax.oss.sink.record.RecordAndStatement;
+import com.datastax.oss.sink.state.InstanceState;
+import com.datastax.oss.sink.state.LifeCycleManager;
+import java.time.Duration;
+import java.time.Instant;
+import java.util.Collection;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.CancellationException;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionStage;
+import java.util.concurrent.ConcurrentLinkedQueue;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.Future;
+import java.util.concurrent.LinkedBlockingQueue;
+import java.util.stream.Collectors;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/** CassandraSinkTask does the heavy lifting of processing records and writing them to DSE. */
+public abstract class RecordProcessor<EngineRecord, EngineHeader> {
```

Comment:
why Header needs to be parameterized? I suspect that both pulsar and Kafka connectors should operate on org.apache.kafka.connect.header.Header isn't it?

Reply:
pulsar sink use its own form of headers. and future sinks for other engines might work with their representations of headers